src/hclust.jl

```diff
@@ -531,6 +531,99 @@ function rorder!(hmer::HclustMerges)
     return hmer
 end
 
+
+## Given a hierarchical cluster and the distance matrix used to generate it,
+## use fast algorithm to determine optimal leaf order minimizing the distance
+## between adjacent leaves. This is done using a heuristic where,
+## when combining multi-leaf sub branches, only the outermost leaves are
+## compared (a maximum of 4 comparisons per intersection).
+##
+## Sub branches are flipped if necessary to minimize the distance between adjacent
+## nodes, and then the combined branches are treated as a block for future
+## comparisons.
+##
+## Based on:
+## [Bar-Joseph et. al. "Fast optimal leaf ordering for hierarchical clustering." _Bioinformatics_. (2001)](https://doi.org/10.1093/bioinformatics/17.suppl_1.S22)
+function orderbranches_barjoseph!(hmer::HclustMerges, dm::AbstractMatrix)
+    order = invperm(hclust_perm(hmer))
+    node_ranges = Tuple{Int,Int}[] # ranges of order array indices occupied by the leaves of each node
+
+    for v in 1:nnodes(hmer)-1
+        vl, vr = hmer.mleft[v], hmer.mright[v]
+
+        (uidx, midx) = node_range(vl, order, node_ranges)
+        (kidx, widx) = node_range(vr, order, node_ranges)
+
+        (u, m) = (order[uidx], order[midx])
+        (k, w) = (order[kidx], order[widx])
+
+        if vl < 0 && vr < 0
+            # Nothing needs to be done
+        elseif vl < 0
+            # check if flipping would reduce distance
+            if dm[m,k] > dm[m,w]
+                reverse!(order, uidx, midx)
+                rotate_merges!(hmer, vr)
+            end
+        elseif vr < 0
+            if dm[k,m] > dm[k,u]
```

Comment:
coveralls reports that this line is not covered by the tests. Do you have some example in mind that would test this line? (Maybe similar to the current band-matrix example, just with different number of elements?)
I still didn't dig into the algorithm details, but I wonder, why both in the `vl < 0` and `vr < 0` branches the same `uidx, midx` region is reversed.

Reply:
>Do you have some example in mind that would test this line?

I could artificially construct some merges that do this, but I think this doesn't actually happen in the wild. I just tried making banded matrices for `n= 2:100`, as well as 1000 random matrices for each of those `n`s and that line is never triggered.

>I still didn't dig into the algorithm details, but I wonder, why both in the vl < 0 and vr < 0 branches the same uidx, midx region is reversed.

That's a mistake in copy-pasting. Should be `kidx` and `widx`

Reply:
See also [this comment](https://github.com/JuliaStats/Clustering.jl/pull/170#discussion_r295945252)

Reply:
Actually, I can't seem to artificially construct the `HclustMerges`... I think [these lines](https://github.com/JuliaStats/Clustering.jl/blob/master/src/hclust.jl#L99-L102) is what's preventing this from happening.

Should I just trust that this won't happen and remove the code? I can comment it out instead in case this ever changes.

Reply:
I think you're right. Could you pls try removing those lines from `push_merges!()`? Now that we support different orderings this flip will just cause confusion.

Reply:
Removing those lines caused 2 test failures:

```
  R hclust() generated examples                                             | 6549         6549
  hclust_n3()                                                               |    1     2      3
```

Reply:
That's no so bad :)
It's the same issue as with `hclust_nn()`.
Could you please wrap `hclust_n3()` call with `Clustering.orderbanches_r!()` in the tests at L84?

Reply:
Done!